client/src/controllers/BlockController.test.js

```diff
@@ -0,0 +1,76 @@
+import { Application } from '@hotwired/stimulus';
+import { BlockController } from './BlockController'; // Replace with your controller path
```

Comment:
Here's an updated starting point for the tests, well done on getting this mostly working, hopefully this helps though.

```js
import { Application } from '@hotwired/stimulus';
import { BlockController } from './BlockController';

const render = jest.fn();
const unpack = jest.fn(() => ({ render }));
window.telepath = { unpack };

describe('BlockController', () => {
  const eventNames = ['w-block:ready'];

  const events = {};

  eventNames.forEach((name) => {
    document.addEventListener(name, (event) => {
      events[name].push(event);
    });
  });

  let application;
  let errors = [];

  const setup = (html, { identifier = 'w-block' } = {}) => {
    document.body.innerHTML = `<main>${html}</main>`;

    application = new Application();

    application.register(identifier, BlockController);

    application.handleError = (error, message) => {
      errors.push({ error, message });
    };

    application.start();

    return Promise.resolve();
  };

  beforeEach(() => {
    application?.stop();
    document.body.innerHTML = '';
    errors = [];
    eventNames.forEach((name) => {
      events[name] = [];
    });
    jest.clearAllMocks();
  });

  it('does nothing if block element is not found', async () => {
    await setup('<div></div>');

    expect(errors).toHaveLength(0);

    expect(unpack).not.toHaveBeenCalled();

    expect(events['w-block:ready']).toHaveLength(0);
  });

  it('should render block if element is controlled', async () => {
    const data = { name: 'John Doe' };

    await setup(
      `<div id="my-element" data-controller="w-block" data-w-block-data-value='${JSON.stringify({ name: 'John Doe' })}'></div>`,
    );

    expect(errors).toHaveLength(0);
    expect(unpack).toHaveBeenCalledWith(data);
    expect(render).toHaveBeenCalledWith(
      document.getElementById('my-element'),
      'my-element',
      [],
      {},
    );
    expect(events['w-block:ready']).toHaveLength(1);
  });

  it('should call the unpacked render function with provided initial & error data', async () => {
    /** */
  });

  it('should throw an error if used on an element without an id', async () => {
    /** */
    // expect(errors).toHaveLength(1);
    // expect(errors).toHaveProperty('0.messages', '???');
  });

  it('should throw an error if Telepath is not available in the window global', async () => {
    /** */
    // expect(errors).toHaveLength(1);
    // expect(errors).toHaveProperty('0.messages', '???');
  });
});
```